### PR TITLE
Huggingface classification support

### PR DESCRIPTION
**Pull request by dirkgr**
This adds support for HF's classification model, similar to the way `HFMC` and `HFQA` work.

### CODE REVIEW

**Comment by jagnusson** [comment]
@dirkgr when I run `python -m catwalk.train --model bert-base-uncased --task mnli` I'm getting a value error inside `collate_for_training()`  in `TrainableHFAutoModel` because it doesn't have a control flow for HF_Classification instance format (just HF_MC and HF_QA).

```python
 def collate_for_training(self, instances: Sequence[Tuple[Task, Instance]]) -> Any:
        mc_instances: List[HFMCInstance] = []
        qa_instances: List[HFQAInstance] = []
        for task, instance in instances:
            if task.has_instance_conversion(InstanceFormat.HF_MC):
                mc_instances.append(cast(HFMCInstance, task.convert_instance(instance, InstanceFormat.HF_MC)))
            elif task.has_instance_conversion(InstanceFormat.HF_QA):
                qa_instances.append(cast(HFQAInstance, task.convert_instance(instance, InstanceFormat.HF_QA)))
            else:
                raise ValueError("I don't know how to handle this instance.")
```

I can dig into this further, but perhaps there's an obvious fix that you can see?

**Comment by dirkgr** [comment]
It's not plumbed through to the training. That's out of scope for this PR.

File: catwalk/models/huggingface.py [hunk 1]
```diff
@@ -116,6 +127,38 @@ def _predict_mc(
                         "acc": (logits, instance.correct_answer_index),
                     }
 
+    def _predict_classification(
```
Comment (jagnusson) [inline catwalk/models/huggingface.py:130]:
```suggestion
    @classmethod
    def _predict_classification(
```

File: catwalk/models/huggingface.py [hunk 2]
```diff
@@ -42,7 +43,17 @@ def predict(  # type: ignore
...
             tokenizer = cached_transformers.get_tokenizer(AutoTokenizer, self.pretrained_model_name_or_path)
             return self._predict_qa(qa_instances, model, tokenizer, batch_size=batch_size)
-        
+        elif task.has_instance_conversion(InstanceFormat.HF_CLASSIFICATION):
+            classification_instances = cast(
+                Sequence[HFClassificationInstance],
+                self._convert_instances(instances, InstanceFormat.HF_CLASSIFICATION, task))
+            model = cached_transformers.get(
+                AutoModelForSequenceClassification,
+                self.pretrained_model_name_or_path, False
```
Comment (jagnusson) [inline catwalk/models/huggingface.py:52]:
```suggestion
                self.pretrained_model_name_or_path, False,
                num_labels=task.num_labels
```

This suggestion needs `Task` to have an attribute saying how many labels it has. Something like this is required to run on anything other than the default `num_labels==2`

File: catwalk/task.py [hunk 3]
```diff
@@ -20,9 +19,6 @@
 
 ENTAILMENT_METRICS = {
     "acc": torchmetrics.Accuracy,
-    "f1": torchmetrics.F1Score,
-    "precision": torchmetrics.Precision,
-    "recall": torchmetrics.Recall
```
Comment (jagnusson) [inline catwalk/task.py:25]: If we make tasks aware of how many classes they have, it would be relatively easy to report F1 as well. That would be desirable right?

File: catwalk/tasks/huggingface.py [hunk 4]
```diff
@@ -138,3 +138,33 @@ def hfmc_conversion(
...
     # We're doing this in this stupid way because this makes the conversion function picklable.
     return functools.partial(hfmc_convert, **kwargs)
+
+
+@dataclass
+class HFClassificationInstance:
+    id: Optional[str]
+    premise: str
+    hypothesis: str
+    label: int
```
Comment (jagnusson) [inline catwalk/tasks/huggingface.py:148]: In my mind nli is only a specific kind of classification. So perhaps we could either make this more general so as to support classification of single or multiple sentence inputs. Or more simply we could just name this HFNLIInstance to avoid confusion.

**Review by jagnusson — CHANGES_REQUESTED** [review]
At present the biggest hang up is just that this code won't run when there are not just 2 classes (such as mnli). This would require making the Task aware of how many classes it has.

More generally, I wonder if it might be best to just call this NLI instead of classification unless we plan to support classification with only a single input sentence as well.

File: catwalk/models/huggingface.py [hunk 5]
```diff
@@ -116,6 +127,38 @@ def _predict_mc(
...
+                    [
+                        (instance.premise, instance.hypothesis)
+                        for instance in batch
+                    ],
+                    padding=True,
+                    truncation="only_first",
+                    return_tensors="pt",
+                    pad_to_multiple_of=8,
+                )
+                results = model(return_dict=True, **tensors)
```
[the remark this answers is not in the document]
Comment (dirkgr) [reply catwalk/models/huggingface.py:155]: How did this ever work without this?

File: catwalk/models/huggingface.py (hunk 2 above)
Comment (dirkgr) [reply catwalk/models/huggingface.py:54]: done

File: catwalk/task.py (hunk 3 above)
Comment (dirkgr) [reply catwalk/task.py]: done

File: catwalk/tasks/huggingface.py (hunk 4 above)
Comment (dirkgr) [reply catwalk/tasks/huggingface.py]: I made it work with n-way classification. Though to really make this useful, we need a way to instantiate an encoder-only model with an arbitrary number of heads. Right now, it will just print a big warning and give you wrong results if you mismatch the labels and the model size.

**Comment by dirkgr** [comment]
@AkshitaB, since @jagnusson is out, can you review this?

File: catwalk/steps.py [hunk 6]
```diff
@@ -199,7 +199,13 @@ def run(
             ]
         dataset_dict = DatasetDict(splits=splits, metadata={})
 
-        trainable_model = model.trainable_copy()
+        trainable_model_kwargs: Dict[str, Any] = {}
+        for task in tasks_in_a_special_variable_because_mypy_is_insane:
+            if isinstance(task, WithAnswerOptionsMixin):
+                trainable_model_kwargs["num_classification_labels"] = max(
+                    trainable_model_kwargs.get("num_classification_labels", 0),
```
Comment (AkshitaB) [inline catwalk/steps.py:206]: When would we want the default number of class labels to be 0?

File: catwalk/tasks/eleuther.py [hunk 7]
```diff
@@ -196,3 +220,20 @@ def get_split(self, split: str) -> Sequence[Dict[str, Any]]:
...
+        answer_options: Sequence[str],
+        version_override: Optional[str] = None,
+    ):
+        EleutherTaskWithRenamedSplits.__init__(
+            self,
+            eleuther_task,
+            version_override=version_override,
+            ranked_classification=True)
+        WithAnswerOptionsMixin.__init__(self, answer_options)
+        self.add_metrics(classification_metrics(len(answer_options)))
```
Comment (AkshitaB) [inline catwalk/tasks/eleuther.py:239]:
```suggestion
        self.add_metrics(classification_metrics(len(answer_options)))

```

**Review by AkshitaB — APPROVED** [review]

File: catwalk/steps.py (hunk 6 above)
Comment (dirkgr) [reply catwalk/steps.py:206]: It's `0` so that the other option in the `max()` function applies, which is `len(task.answer_options)`.
